### PR TITLE
Paas - New-CmAzPaasSharedImageGallery initial draft

### PR DESCRIPTION
**Pull request by deepend-dev**

### CODE REVIEW

File: Cloudmarque.Azure/Public/PaaS/sharedImageGallery/New-CmAzpaasSharedImageGallery.json [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+                "description": "Private Gallery Created by Cloudmarque"
+            }
+        },
+        {
+            "type": "Microsoft.Resources/Deployments",
+            "apiVersion": "2019-10-01",
+            "name": "[concat('Cm_', parameters('GalleryName'), '_', parameters('SIGObjectArray')[copyIndex('SIGObjectArrayCopy')].definitionName)]",
+            "copy": {
+                "name": "SIGObjectArrayCopy",
+                "count": "[length(parameters('SIGObjectArray'))]"
```
Comment (tm-luke) [inline Cloudmarque.Azure/Public/PaaS/sharedImageGallery/New-CmAzpaasSharedImageGallery.json:38]: Can we simplify this name? If its a list of shared image galleries then SharedImageGalleryCopy is fine.

File: Cloudmarque.Azure/Public/PaaS/sharedImageGallery/New-CmAzpaasSharedImageGallery.json [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+            },
+            "dependsOn": [
+                "[parameters('GalleryName')]"
+            ],
+            "properties": {
+                "mode": "Incremental",
+                "expressionEvaluationOptions": {
+                    "scope": "inner"
+                },
```
Comment (tm-luke) [inline Cloudmarque.Azure/Public/PaaS/sharedImageGallery/New-CmAzpaasSharedImageGallery.json:47]: Can we use outer here? Means less parameter copying

File: Cloudmarque.Azure/Public/PaaS/sharedImageGallery/New-CmAzpaasSharedImageGallery.ps1 [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+                }
+                else {
+                    $date = ((Get-date).AddYears('5').GetDateTimeFormats('s')).split()
+
+                    Write-Verbose "Default: End of life date will be set to : $date"
+                    $_.endOfLifeDate = $date
+                }
+
+                if (!$_.location) {
+                    Write-Verbose
```
Comment (tm-luke) [inline Cloudmarque.Azure/Public/PaaS/sharedImageGallery/New-CmAzpaasSharedImageGallery.ps1:83]: Unsure about this

File: Cloudmarque.Azure/Public/PaaS/sharedImageGallery/New-CmAzpaasSharedImageGallery.ps1 [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+            elseif (!$resourceGroup -and !$SettingsObject.resourceGroupName) {
+                Write-Error "Please provide appropriate service tag for existing resource group or provide unique name to create new."
+            }
+
+            $SettingsObject.imageDefinations | ForEach-Object {
+
+                Write-Verbose "Validating ImageDefination: $($_.definitionName)"
+
+                if ($_.endOfLifeDate) {
+                    $_.endOfLifeDate = (([datetime]::Parse($_.endOfLifeDate)).GetDateTimeFormats('s')).Split()
```
Comment (tm-luke) [inline Cloudmarque.Azure/Public/PaaS/sharedImageGallery/New-CmAzpaasSharedImageGallery.ps1:73]: Powershell-yaml handles datetimes, do we need the parsing here?

File: Cloudmarque.Azure/Public/PaaS/sharedImageGallery/New-CmAzpaasSharedImageGallery.ps1 [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+                }
+                else {
+                    $date = ((Get-date).AddYears('5').GetDateTimeFormats('s')).split()
```
Comment (tm-luke) [inline Cloudmarque.Azure/Public/PaaS/sharedImageGallery/New-CmAzpaasSharedImageGallery.ps1:76]: Wondering if we can set this to the datetime maxvalue, not filling in a endOfLifeDate implies the user doesn't want it to expire.

File: Cloudmarque.Azure/Public/PaaS/sharedImageGallery/New-CmAzpaasSharedImageGallery.ps1 [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+                    Write-Verbose "Default: End of life date will be set to : $date"
+                    $_.endOfLifeDate = $date
+                }
+
+                if (!$_.location) {
+                    Write-Verbose
+                    $_.location = $SettingsObject.location
+                }
+
+                if (!$_.osType) {
```
Comment (tm-luke) [inline Cloudmarque.Azure/Public/PaaS/sharedImageGallery/New-CmAzpaasSharedImageGallery.ps1:87]: Maybe set the variable, then add it to the verbose string, apply the same to any similar instances below.

File: Cloudmarque.Azure/Public/PaaS/sharedImageGallery/New-CmAzpaasSharedImageGallery.ps1 [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+                    }
+
+                }
+
+                New-AzResourceGroupDeployment `
+                    -Name "Cm_Gallery_$($SettingsObject.galleryName)" `
+                    -ResourceGroupName $resourceGroup.ResourceGroupName `
+                    -TemplateFile "$PSScriptRoot\New-CmAzpaasSharedImageGallery.json" `
+                    -Location $SettingsObject.location `
+                    -SIGObjectArray $SettingsObject.imageDefinations `
```
Comment (tm-luke) [inline Cloudmarque.Azure/Public/PaaS/sharedImageGallery/New-CmAzpaasSharedImageGallery.ps1:139]: imageDefinitions

File: Cloudmarque.Azure/Public/PaaS/sharedImageGallery/sharedImageGallery.yml [hunk 8]
```diff
@@ -0,0 +1,21 @@
+resourceGroupServiceTag:
+location:
+
+galleryName:
+
+imageDefinations:
+  - definitionName:
+    publisher: "string"
```
Comment (tm-luke) [inline Cloudmarque.Azure/Public/PaaS/sharedImageGallery/sharedImageGallery.yml:8]: Tab out any comments in settings.

**Review by tm-luke — CHANGES_REQUESTED** [review]

File: Cloudmarque.Azure/Public/PaaS/sharedImageGallery/New-CmAzpaasSharedImageGallery.json (hunk 2 above)
Comment (deepend-dev) [reply Cloudmarque.Azure/Public/PaaS/sharedImageGallery/New-CmAzpaasSharedImageGallery.json]: there are multiple levels of inner loops which wont work

File: Cloudmarque.Azure/Public/PaaS/sharedImageGallery/New-CmAzpaasSharedImageGallery.json (hunk 1 above)
Comment (deepend-dev) [reply Cloudmarque.Azure/Public/PaaS/sharedImageGallery/New-CmAzpaasSharedImageGallery.json]: done

File: Cloudmarque.Azure/Public/PaaS/sharedImageGallery/New-CmAzpaasSharedImageGallery.ps1 (hunk 3 above)
Comment (deepend-dev) [reply Cloudmarque.Azure/Public/PaaS/sharedImageGallery/New-CmAzpaasSharedImageGallery.ps1]: thanks for catching this. corrected

File: Cloudmarque.Azure/Public/PaaS/sharedImageGallery/New-CmAzpaasSharedImageGallery.ps1 (hunk 5 above)
Comment (deepend-dev) [reply Cloudmarque.Azure/Public/PaaS/sharedImageGallery/New-CmAzpaasSharedImageGallery.ps1]: ideally they should. as there will be security fix and vulnerability fixes..

File: Cloudmarque.Azure/Public/PaaS/sharedImageGallery/New-CmAzpaasSharedImageGallery.ps1 (hunk 6 above)
Comment (deepend-dev) [reply Cloudmarque.Azure/Public/PaaS/sharedImageGallery/New-CmAzpaasSharedImageGallery.ps1]: didnt get this

File: Cloudmarque.Azure/Public/PaaS/sharedImageGallery/New-CmAzpaasSharedImageGallery.ps1 (hunk 7 above)
Comment (deepend-dev) [reply Cloudmarque.Azure/Public/PaaS/sharedImageGallery/New-CmAzpaasSharedImageGallery.ps1]: done

File: Cloudmarque.Azure/Public/PaaS/sharedImageGallery/sharedImageGallery.yml (hunk 8 above)
Comment (deepend-dev) [reply Cloudmarque.Azure/Public/PaaS/sharedImageGallery/sharedImageGallery.yml]: done

File: Cloudmarque.Azure/Public/PaaS/sharedImageGallery/New-CmAzPaasSharedImageGallery.ps1 [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+                }
+                else {
+                    $date = ((Get-date).AddYears('5').GetDateTimeFormats('s')).split()
+
+                    Write-Verbose "Default: End of life date will be set to : $date"
+                    $_.endOfLifeDate = $date
+                }
+
+                if (!$_.location) {
+                    Write-Verbose "Default: image defination location will be defaulted Shared imaged gallery : $($SettingsObject.location)"
```
Comment (tm-luke) [inline Cloudmarque.Azure/Public/PaaS/sharedImageGallery/New-CmAzPaasSharedImageGallery.ps1:83]: definition not defination

Comment (tm-luke) [reply Cloudmarque.Azure/Public/PaaS/sharedImageGallery/New-CmAzPaasSharedImageGallery.ps1:83]: There seem to be multiple instances of this, can you change this in all files please

File: Cloudmarque.Azure/Public/PaaS/sharedImageGallery/New-CmAzPaasSharedImageGallery.ps1 [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+                    $_.location = $SettingsObject.location
+                }
+
+                if (!$_.osType) {
+                    Write-Verbose "Default: OS type will be set to - Linux"
+                    $_.osType = "Linux"
```
Comment (tm-luke) [inline Cloudmarque.Azure/Public/PaaS/sharedImageGallery/New-CmAzPaasSharedImageGallery.ps1:89]: What i mean is, maybe set the ostype, the use it in the string, theres a few places where this change can be applied.

File: Cloudmarque.Azure/Public/PaaS/sharedImageGallery/New-CmAzPaasSharedImageGallery.ps1 [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+
+            $SettingsObject.imageDefinations | ForEach-Object {
+
+                Write-Verbose "Validating ImageDefination: $($_.definitionName)"
+
+                if ($_.endOfLifeDate) {
+                    $_.endOfLifeDate = (([datetime]::Parse($_.endOfLifeDate)).GetDateTimeFormats('s')).Split()
+                }
+                else {
+                    $date = ((Get-date).AddYears('5').GetDateTimeFormats('s')).split()
```
Comment (tm-luke) [inline Cloudmarque.Azure/Public/PaaS/sharedImageGallery/New-CmAzPaasSharedImageGallery.ps1:76]: Good point, adding a few years is fine actually.

**Review by tm-luke — CHANGES_REQUESTED** [review]

File: Cloudmarque.Azure/Public/PaaS/sharedImageGallery/New-CmAzPaasSharedImageGallery.ps1 (hunk 9 above)
Comment (deepend-dev) [reply Cloudmarque.Azure/Public/PaaS/sharedImageGallery/New-CmAzPaasSharedImageGallery.ps1:83]: yes 👍

**Review by tm-luke — DISMISSED** [review]

**Review by tm-luke — APPROVED** [review]

**Pull request merged**